reactor-core/src/main/java/reactor/core/publisher/FluxBufferUntilChanged.java

```diff
@@ -0,0 +1,322 @@
+						"requested");
+
+		static final AtomicReferenceFieldUpdater<BufferUntilChangedSubscriber, Subscription>
+				S =
+				AtomicReferenceFieldUpdater.newUpdater(BufferUntilChangedSubscriber.class,
+						Subscription.class,
+						"s");
+
+		@Nullable
+		K lastKey;
+
+		public BufferUntilChangedSubscriber(CoreSubscriber<? super C> actual,
+				Function<? super T, K> keyExtractor,
+				BiPredicate<? super K, ? super K> keyComparator,
+				C initialBuffer,
+				Supplier<C> bufferSupplier) {
+			this.actual = actual;
+			this.keyExtractor = keyExtractor;
+			this.keyComparator = keyComparator;
+			this.buffer = initialBuffer;
+			this.bufferSupplier = bufferSupplier;
+		}
+
+		@Override
+		public CoreSubscriber<? super C> actual() {
+			return actual;
+		}
+
+		@Override
+		public void onSubscribe(Subscription s) {
+			if (Operators.setOnce(S, this, s)) {
+				actual.onSubscribe(this);
+			}
+		}
+
+		@Override
+		public void onNext(T t) {
+			if (!tryOnNext(t)) {
+				s.request(1);
+			}
+		}
+
+		@Override
+		public void onError(Throwable t) {
+			if (done) {
+				Operators.onErrorDropped(t, actual.currentContext());
+				return;
+			}
+			done = true;
+			Operators.onDiscardMultiple(buffer, actual.currentContext());
+			buffer = null;
+			lastKey = null;
+			actual.onError(t);
+		}
+
+		@Override
+		public void onComplete() {
+			if (done) {
+				return;
+			}
+			done = true;
+			lastKey = null;
+			DrainUtils.postComplete(actual, this, REQUESTED, this, this);
+		}
+
+		@Override
+		public void request(long n) {
+			if (Operators.validate(n)) {
+				if (n == Long.MAX_VALUE) {
+					fastpath = true;
+					requested = Long.MAX_VALUE;
+					s.request(Long.MAX_VALUE);
+				}
+				else if (!DrainUtils.postCompleteRequest(n,
+						actual,
+						this,
+						REQUESTED,
+						this,
+						this)) {
+					s.request(1);
+				}
+			}
+		}
+
+		@Override
+		public void cancel() {
+			Operators.terminate(S, this);
+			Operators.onDiscardMultiple(buffer, actual.currentContext());
+			lastKey = null;
+		}
+
+		@Override
+		public boolean tryOnNext(T t) {
+			if (done) {
+				Operators.onNextDropped(t, actual.currentContext());
+				return true;
+			}
+
+			C b = buffer;
+			K k;
+
+			try {
+				k = Objects.requireNonNull(keyExtractor.apply(t),
+						"The distinct extractor returned a null value.");
+			}
+			catch (NullPointerException e) {
+				Context ctx = actual.currentContext();
+				onError(Operators.onOperatorError(s, e, t, ctx));
+				Operators.onDiscardMultiple(buffer, ctx);
+				Operators.onDiscard(t, ctx);
+				return true;
+			}
+
+			if (null == lastKey) {
+				lastKey = k;
+				b.add(t);
+				return true;
+			}
+
+			boolean match;
+
+			match = keyComparator.test(lastKey, k);
+			lastKey = k;
+
+			boolean requestMore;
+			if (match) {
+				b.add(t);
+			}
+			else {
+				requestMore = onNextNewBuffer();
+				b = buffer;
+				b.add(t);
+				return !requestMore;
+			}
+
+			return !(!fastpath && requested != 0);
```

Comment:
> considering how close this is to `FluxBufferPredicate`, I'm wondering if a version based off a special `Predicate` built from the user-provided `BiPredicate` and `keyExtractor`, which maintains the `lastKey` state, would work. Goal would be to reuse most of the existing operator's code. Only change it would need is some overridable "cleanup" method invoked in onComplete/onError/cancel to set the `lastKey` state to `null`. (or detect that the stateful `Predicate` is a `Disposable` and let it clean itself up in these same cases). wdyt @UgiR ?

Sounds good.